Cloudwatch: Add support for resolution of resource tags in alias

Alex-Richman opened the pull request:
This PR adds support for using resource tags in the CloudWatch query alias.  This is quite useful since the CloudWatch metric label for queries with a resource dimension is the resource ID, which is not particularly human-readable.

Seems to be a fairly desired feature [0][1][2][3][4][5], and personally I couldn't see myself using the CloudWatch datasource without it (who remembers their CloudFront distribution IDs?).

Existing "solutions" (read: bodges) involve using dashboard variable queries (ec2_instance_attributes) and value groups[3], which prevents use of Alerts and only supports EC2 instances.

Resources are fetched using the `resourcegroupstaggingapi` as appropriate for the Namespace in the queries.  For example, a query in the `AWS/EC2` namespace/dimension fetches all resources of type `ec2:instance` and populates a map with their tags.  If all queries are in the same region then this is a single RGTA request for all the queries, so the performance impact is negligible.

The fetched resource tag map is then injected into the `formatAlias` data map based on the CloudWatch metric label, so that tags can be used in the query alias string.

Currently supports fetching resources for EC2 (InstanceID), CloudFront (DistributionId), ApplicationELB (LoadBalancer, TargetGroup), ECS (ClusterName, ServiceName), RDS (DBInstanceIdentifier), and S3 (BucketName).  More resources can be easily added by updating the `resourceTypesMap`, but the current set is all I'm using right now so I wouldn't have any resources to test against for the other types.

Note also that CloudWatch is somewhat inconsistent in how it constructs the metric labels, so we have to do a little feathering to make the label components consistent with the resource IDs from RGTA.  For example ELBv2 TargetGroups have an extra `targetgroup/` prefix in the CloudWatch label compared to the RGTA resource ID.

Closes #27280

[0] https://community.grafana.com/t/cloudwatch-show-servername-instead-of-instance-id/10962
[1] https://community.grafana.com/t/multiple-ec2-instances-in-grafana-dashboard/22032
[2] https://github.com/monitoringartist/grafana-aws-cloudwatch-dashboards/issues/19
[3] Comments in https://github.com/grafana/grafana/pull/8307
[4] https://github.com/grafana/grafana/issues/3478
[5] https://github.com/grafana/grafana/issues/7294

Alex-Richman committed:
Cloudwatch: Add support for resolution of resource tags in alias

- Add RGTA resolution of resource tags involved in the query (per Namespace/Dimensions)
- Inject tags into formatAlias dictionary to allow use in aliases

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=27281) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=27281) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=27281) it.</sub>

Alex-Richman committed:
Unbreak CloudWatch response_parser tests

Alex-Richman committed:
Cloudwatch: Optimise resource tag fetching

- Fetch all queries' resource types in a single RGTA query (split per region)
- Select resource types per Namespace not Namespace/Dimension so that custom expressions can use tags

aknuds1 commented on pkg/tsdb/cloudwatch/response_parser.go:
```diff
@@ -7,11 +7,91 @@ import (
+	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/cloudwatch"
+	"github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
+	"github.com/gigawattio/awsarn"
```
This adds a dependency that should be reflected in go.mod/go.sum.

aknuds1 reviewed (changes requested):
Please see comments.

Alex-Richman committed:
Apply suggestions from code review

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

Alex-Richman committed:
Cloudwatch: Further fixup after applying PR suggestions

Alex-Richman commented:
> Please see comments.

Thanks for your review - I've applied your suggestions and made a further fixup commit to take care of the rest.  Let me know if there's anything else.

aknuds1 commented:
Thanks @Alex-Richman, looks better now, but I'll have to read more to get a grip on the functionality.

aknuds1 commented on pkg/tsdb/cloudwatch/response_parser.go:
```diff
@@ -190,6 +275,19 @@ func formatAlias(query *cloudWatchQuery, stat string, dimensions map[string]stri
+	/* CloudWatch labels for resource dimensions are space separated lists of the resource IDs.
+	 * A caller might want to access tags for any of the resources in the label, so we inject the tags with a unique prefix per part.
+	 */
+	for i, part := range strings.Split(label, " ") {
+		if _, ok := resourceTags[part]; !ok {
+			continue
+		}
+
+		for key, value := range resourceTags[part] {
+			data[fmt.Sprintf("tags.%d.%s", i, key)] = value
+		}
+	}
```
It's more understandable to me if the variable is named `resource`, since it represents a resource ID. It's also more robust to use the default value in `resourceTags[resource]`, since it will correspond to an empty map:
```suggestion
	for i, resource := range strings.Split(label, " ") {
		tags := resourceTags[resource]
		for key, value := range tags {
			data[fmt.Sprintf("tags.%d.%s", i, key)] = value
		}
	}
```

aknuds1 reviewed (dismissed):
Left more suggestions.

I haven't 100% got my head around the functionality yet, but I think we might need tests for `parseResponse` that exercise the new functionality.

Alex-Richman committed:
Apply suggestions from code review

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

Alex-Richman committed:
Cloudwatch: Further fixup after applying PR suggestions (round 2)

Alex-Richman commented:
@aknuds1 Applied and resolved round 2, thanks again

Alex-Richman committed:
Clouwatch: Add response parser resource tag expansion tests

Alex-Richman commented:
@aknuds1 Added a few tests related to the functionality

aknuds1 commented:
Thanks - I will look more at this when time allows.

aknuds1 reviewed (commented):
Just a nit I spotted.

Alex-Richman committed:
Apply suggestions from code review

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

Alex-Richman commented:
@aknuds1 Thanks - Applied.

Easiest way to test would be to spin up a couple of EC2 instances, add some tags to them, configure a panel with a CloudWatch datasource query for an EC2 metric (for example CPUUtilization) and play around with the Alias field to render tags from the instances.

Here's a sample panel before & after using a tag based alias (excuse the redactions).
![1](https://user-images.githubusercontent.com/1280519/92748817-705fa780-f37d-11ea-8ced-f849c7f3d026.png)
![2](https://user-images.githubusercontent.com/1280519/92748814-6fc71100-f37d-11ea-8800-b881d501aa35.png)

marefr commented on pkg/tsdb/cloudwatch/response_parser.go:
```diff
@@ -7,11 +7,85 @@ import (
+			}
+		}
+
+		resources, err := e.resourceGroupsGetResources(region, nil, typeFilters)
```
What's the performance/cost of doing this call per each query/region? At a first glance seems like it could be problematic?

Alex-Richman replied:
Queries in the same region are consolidated into the same RGTA query, so for most panels (which are single region) this will just be a single RGTA call.

RGTA is a very lightweight API (basically just key/value fetching), so the performance impact is negligible.

Additionally, the existing "solution" for this sort of lookup (at least for EC2 instances) is via ec2_instance_attribute at the dashboard level with query variables, which just boils down to a bunch of RGTA queries in the background anyway (one per instance, rather than one per panel).

marefr commented:
@Alex-Richman thanks for the contribution. However, the referenced feature request #27280 haven't seen any activity/comments/upvotes, that together with the extra complexity, maintenance burden and possibly cost (performance and calls) makes it hard for us to accept this currently. We expect CloudWatch API to have this support built-in when querying metrics. Obviously it doesn't, but currently don't think Grafana should try to solve this limitation. We're very sorry we've given you false expectations accepting this change by requesting changes and so on.  

I'll close this now since we're trying to only have relevant/acceptable PR's open, but please note that a closed PR can always be reopened should a feature request see more activity/discussions or if I misunderstood something. Would suggest to continue the discussions, if any, in the feature request issue #27280.

papagian commented:
Adding a comment for future reference:
this feature only makes sense if all the timeseries have the specific tag defined
otherwise the other timeseries are not interpolated:
![Screenshot 2020-09-16 at 17 06 55](https://user-images.githubusercontent.com/1632407/94015728-ec161700-fdb5-11ea-831a-ceea0ae9133e.png)

waqark3389 commented:
@papagian Well obviously... 

@marefr Its very annoying that you have closed this. This has driven me to not buy grafana enterprise after much consideration and move to Datadog.
Especially since @Alex-Richman took the time to fix and make improvments on suggestions, you go ahead and close it.

marefr commented:
@waqark3389 I think I was quite detailed in my response to why we decided to close it and it doesn't mean we can't reopen it. Please provide more information in #27280 to help us understand the problem better and to give other users higher chances of finding this issue by linking all related information you mention together:

> you state you havent seen any activity etc. This question on tags (specifically instance name) has come up countless times on the forum, in IRC, in Slack and in previous github issues. I hadnt upvoted the feature request because I hadnt seen it, as I'm sure is the case with many others.

Thanks

wallet72 commented:
Is there a way to make this an option or a module? I see the cost and workload argument but to be honest if I have to display instance IDs and not tag.Name then I'm going to have a lot of very confused managers.

I mean if you are not familiar with AWS, what is easier to interpret - a name or an instance ID?

Considering these guys pay the bills, they won't like paying for Grafana if it just confuses them.

macobo commented:
Since more voices were asked for - this is a must-have feature. EC2 is the backbone of aws. Cloudwatch is a great way to get metrics out for EC2 into grafana. 

However without usable labels what use is there for metrics?

chrisolido commented:
I believe we really need to push this tags feature as all resources are being tag. Its a must for diplaying human readable Name than label we see now.
![image](https://user-images.githubusercontent.com/343307/121843997-d6ff6080-cd15-11eb-88db-ee64ae7f1e2d.png)

mean not working 
![image](https://user-images.githubusercontent.com/343307/121844071-f26a6b80-cd15-11eb-8356-6b64b8ca8cdf.png)

{{tags.0.Name}} is not working
![image](https://user-images.githubusercontent.com/343307/121844202-32c9e980-cd16-11eb-850b-060946bb7046.png)

kmoberg commented:
Agreed with previous statements here: this should be a no-brainer issue to fix since the fix has already been created and is almost ready to implement. AWS clearly isn't going to change, and we have the opportunity to solve the entire issue by implementing this PR. 

@marefr  Please reconsider this, as others have stated - managers that pay the bills can't have instance IDs shown but need instance names, and us not being able to show this, most likely will force us to move to Datadog or another service other than Grafana Enterprise.

Seven2Nine commented:
![image](https://user-images.githubusercontent.com/61679388/194760674-60b233d0-d84e-4dda-8b23-9b8be9e622aa.png)
why latest grafana version don't have alias field.

marefr commented:
@grafana/aws-plugins can you please have a look at comment above? Seems like a potential bug

fridgepoet commented:
Hi @Seven2Nine, Alias has been replaced by the `Label` field which supports [Dynamic Labels].(https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/graph-dynamic-labels.html)
See more about that change here -- https://github.com/grafana/grafana/issues/48434

sunker commented:
Looks like the Monaco editor in the `Label` field is not being rendered properly here. @Seven2Nine would you mind opening the javascript console and checking if there are any errors? Also, what browser are you using?
